Compute type_equal() without reference to backend()

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #54079 Add Tensor::is_cpu, genericize TensorIterator
* #54034 Refactor tensor_new.cpp to use TensorOptions instead of DispatchKey
* #54016 Delete denseTypeIdWithDefault and toDense
* #54013 Delete defunct ComplexCPU/ComplexCUDA dispatch keys
* #53972 Make storage access error NotImplementedError
* **#53823 Compute type_equal() without reference to backend()**
* #53816 Also propagate storage_access_should_throw_ when copying tensor metadata

Argument for correctness: type_equal previous compared if backends
are equal.  Backend is computed by translation from dispatch key.
I verified that computeDispatchKey never computed a weird
dispatch key (e.g., AutogradXLA), so that dispatchKeyToBackend
was effectively injective.  Then it is always valid to compare
the arguments of an injective function for equality, rather than
the output of the injective function.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D27036575](https://our.internmc.facebook.com/intern/diff/D27036575)

ezyang committed:
Compute type_equal() without reference to backend()

Argument for correctness: type_equal previous compared if backends
are equal.  Backend is computed by translation from dispatch key.
I verified that computeDispatchKey never computed a weird
dispatch key (e.g., AutogradXLA), so that dispatchKeyToBackend
was effectively injective.  Then it is always valid to compare
the arguments of an injective function for equality, rather than
the output of the injective function.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

bhosmer reviewed (approved):
This makes sense but it seems pretty fragile. 
- any idea what it saves us, perf-wise?
- can we throw in a lightweight injectivity test somewhere?

ezyang commented:
> any idea what it saves us, perf-wise?

My intention wasn't really perf; I just really didn't want to put Meta in the list of Backends XD

> can we throw in a lightweight injectivity test somewhere?

I... guess? But honestly I'd rather spend the time just getting rid of Backend (as we have kept saying we would do, and then haven't)

bhosmer commented:
> I... guess? But honestly I'd rather spend the time just getting rid of Backend (as we have kept saying we would do, and then haven't)

Makes sense. So maybe just a comment on `computeDispatchKey()` noting that we use its output as a proxy for backend so it must only return keys for which `dispatchKeyToBackend()` is injective

ezyang committed:
Update on "Compute type_equal() without reference to backend()"


Argument for correctness: type_equal previous compared if backends
are equal.  Backend is computed by translation from dispatch key.
I verified that computeDispatchKey never computed a weird
dispatch key (e.g., AutogradXLA), so that dispatchKeyToBackend
was effectively injective.  Then it is always valid to compare
the arguments of an injective function for equality, rather than
the output of the injective function.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Compute type_equal() without reference to backend()"


Argument for correctness: type_equal previous compared if backends
are equal.  Backend is computed by translation from dispatch key.
I verified that computeDispatchKey never computed a weird
dispatch key (e.g., AutogradXLA), so that dispatchKeyToBackend
was effectively injective.  Then it is always valid to compare
the arguments of an injective function for equality, rather than
the output of the injective function.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D27036575](https://our.internmc.facebook.com/intern/diff/D27036575)

[ghstack-poisoned]

ezyang committed:
Update on "Compute type_equal() without reference to backend()"


Argument for correctness: type_equal previous compared if backends
are equal.  Backend is computed by translation from dispatch key.
I verified that computeDispatchKey never computed a weird
dispatch key (e.g., AutogradXLA), so that dispatchKeyToBackend
was effectively injective.  Then it is always valid to compare
the arguments of an injective function for equality, rather than
the output of the injective function.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D27036575](https://our.internmc.facebook.com/intern/diff/D27036575)

[ghstack-poisoned]

ezyang committed:
Update on "Compute type_equal() without reference to backend()"


Argument for correctness: type_equal previous compared if backends
are equal.  Backend is computed by translation from dispatch key.
I verified that computeDispatchKey never computed a weird
dispatch key (e.g., AutogradXLA), so that dispatchKeyToBackend
was effectively injective.  Then it is always valid to compare
the arguments of an injective function for equality, rather than
the output of the injective function.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D27036575](https://our.internmc.facebook.com/intern/diff/D27036575)

[ghstack-poisoned]

ezyang committed:
Update on "Compute type_equal() without reference to backend()"


Argument for correctness: type_equal previous compared if backends
are equal.  Backend is computed by translation from dispatch key.
I verified that computeDispatchKey never computed a weird
dispatch key (e.g., AutogradXLA), so that dispatchKeyToBackend
was effectively injective.  Then it is always valid to compare
the arguments of an injective function for equality, rather than
the output of the injective function.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D27036575](https://our.internmc.facebook.com/intern/diff/D27036575)

[ghstack-poisoned]

ezyang committed:
Update on "Compute type_equal() without reference to backend()"


Argument for correctness: type_equal previous compared if backends
are equal.  Backend is computed by translation from dispatch key.
I verified that computeDispatchKey never computed a weird
dispatch key (e.g., AutogradXLA), so that dispatchKeyToBackend
was effectively injective.  Then it is always valid to compare
the arguments of an injective function for equality, rather than
the output of the injective function.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D27036575](https://our.internmc.facebook.com/intern/diff/D27036575)

[ghstack-poisoned]
